nixos/modules/services/audio/snapserver.nix

```diff
@@ -0,0 +1,219 @@
+        type = types.ints.u16;
+        default = 1705;
+        description = ''
+          The port for control connections (JSON-RPC).
+        '';
+      };
+
+      openFirewall = mkOption {
+        type = types.bool;
+        default = true;
+        description = ''
+          Whether to automatically open the specified ports in the firewall.
+        '';
+      };
+
+      inherit sampleFormat;
+      inherit codec;
+      inherit streamBuffer;
+
+      streams = mkOption {
+        type = with types; attrsOf (submodule {
+          options = {
+            location = mkOption {
+              type = types.path;
+              default = /run + "/${name}/snapfifo";
+              description = ''
+                The location of the pipe.
+              '';
+            };
+            type = mkOption {
+              type = types.enum [ "pipe" "file" "process" "spotify" "airplay" ];
+              default = "pipe";
+              description = ''
+                The type of input stream.
+              '';
+            };
+            mode = mkOption {
+              type = with types; nullOr (enum [ "create" "read" ]);
+              default = null;
+              example = "read";
+              description = ''
+                The opening mode of the object to read from.
+                By default, snapcast creates a pipe or file if it
+                doesn't exist. If an audio source process requires
+                to create the file itself, it can be set to read.
+              '';
+            };
+            inherit sampleFormat;
+            inherit codec;
+            host = streamHost;
+            buffer = streamBuffer;
+          };
+        });
+        default = {
+          default = {
+            location = location.default;
+            host = streamHost.default;
+            codec = codec.default;
+            buffer = streamBuffer.default;
+            sampleFormat = sampleFormat.default;
+          };
+        };
+      };
+
+      buffer = mkOption {
+        type = with types; nullOr int;
+        default = null;
+        description = ''
+          Buffer in ms.
+        '';
+      };
+
+      sendToMuted = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Send audio to muted clients.
+        '';
+      };
+
+      dataDir = mkOption {
+        type = types.path;
+        default = /var/lib + "/${name}";
+        description = ''
+          The directory where snapserver stores its state.
+        '';
+      };
+    };
+
+  };
+
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+
+    systemd.services.snapserver = {
+      after = [ "network.target" ];
+      description = "Snapserver";
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        PermissionsStartOnly = true;
+        DynamicUser = true;
+        ExecStart = "${pkgs.snapcast}/bin/snapserver --daemon ${optionString}";
+        Type = "forking";
```

Comment:
For correctness, this has to be "forking" unfortunately. "exec" would start dependent services too early and "notify" is not supported by `snapserver`. Socket activation is difficult to implement because of the multiple ports and pipes the server opens / listens on.

Reply:
> "exec" would start dependent services too early

What do you mean by that? What's "exec" and what's "notify"?

Reply:
Other possible types of services. See https://www.freedesktop.org/software/systemd/man/systemd.service.html#Options.

Reply:
Ohh I see, so systemd considers a forking service to be in the starting phase when it hasn't forked yet, and delays dependencies until it forks. Whereas with `exec`, it considers it started right away. That correct?

Reply:
As far as I understand yes.
BTW that brings me to anther question: Should potentially dependent services like mpd and mopidy be put in the `Before` list? I did not do this initially because the module does not know if those are actually configured to output to snapcast if enabled. But I guess that this is usually what people would want. What do you think?

Reply:
You should probably put this in a section conditioned on both services being enabled.

Reply:
Systemd already has that condition internally, I don't see a reason to duplicate that logic in the modules.